File: desktopclient/src/main/scala/com/goexp/galgame/gui/util/CommonTabController.java
package com.goexp.galgame.gui.util;

import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.view.game.listview.DataViewController;
import javafx.beans.value.ChangeListener;
import javafx.collections.ObservableList;
import javafx.concurrent.Service;
import javafx.concurrent.Task;
import javafx.scene.layout.Region;

import java.util.function.Supplier;

public class CommonTabController {

    public Region node;
    public DataViewController controller;

    private Service<ObservableList<Game>> gameSearchService;


    public CommonTabController(Supplier<Task<ObservableList<Game>>> task) {
        this.gameSearchService = new TaskService<>(task);

        init();
    }

    public CommonTabController(Service<ObservableList<Game>> service) {
        this.gameSearchService = service;

        init();
    }

    private void init() {

        final var loader = new FXMLLoaderProxy<Region, DataViewController>("view/game_explorer/dataview.fxml");
        node = loader.node;
        controller = loader.controller;

        controller.reloadProperty.addListener((observable, oldValue, newValue) -> {
            if (newValue != null && newValue)
                gameSearchService.restart();
        });

        gameSearchService.valueProperty().addListener((observable, oldValue, newValue) -> {
            if (newValue != null) {
                controller.setItems(newValue);
            }
        });

        controller.progessloading.visibleProperty().bind(gameSearchService.runningProperty());

    }

    public void load() {
        gameSearchService.restart();
    }

}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/game/CommonTabController.scala
package com.goexp.galgame.gui.view.game

import com.goexp.galgame.gui.model.Game
import com.goexp.galgame.gui.util.SimpleFxmlLoader
import com.goexp.galgame.gui.view.game.listview.DataViewController
import com.goexp.ui.javafx.TaskService
import javafx.collections.ObservableList
import javafx.concurrent.Task

class CommonTabController(private val taskCreator: () => Task[ObservableList[Game]]) {

  private val loader = new SimpleFxmlLoader[DataViewController]("dataview.fxml")
  val node = loader.node
  val controller = loader.controller

  private val queryService = new TaskService(taskCreator)

  init()

  private def init() = {

    queryService.valueProperty.addListener((_, _, newValue) => {
      if (newValue != null) {
        controller.load(newValue)
        //        controller.loadingBar.setVisible(false)
      }
    })

    //    controller.loadingBar.visibleProperty.bind(queryService.runningProperty)
  }

  def load(): Unit = {
    queryService.restart()
    //    controller.loadingBar.setVisible(true)
  }


}

object CommonTabController {
  def apply(taskCreator: => Task[ObservableList[Game]]): CommonTabController = new CommonTabController(() => taskCreator)
}